physicalplan/function/function.go

```diff
@@ -1,101 +1,279 @@
+		return ignoreEmptyScalar(f.Points, f.ScalarPoints, calculatedSample(f.Labels, f.StepTime, f.Points, f.ScalarPoints, clampMin))
+	},
+	"clamp_max": func(f FunctionArgs) promql.Sample {
+		return ignoreEmptyScalar(f.Points, f.ScalarPoints, calculatedSample(f.Labels, f.StepTime, f.Points, f.ScalarPoints, clampMax))
+	},
+}
+
+// functionSelector returns []model.StepVector after processing input with desired function.
+type functionSelector struct {
+	vectorPool    *model.VectorPool
+	funcExpr      *parser.Call
+	mint          int64
+	maxt          int64
+	step          int64
+	currentStep   int64
+	stepsBatch    int
+	series        []labels.Labels
+	once          sync.Once
+	majorArgIndex int
+	majorArg      model.VectorOperator
+	scalarArgs    []model.VectorOperator
+
+	call FunctionCall
+}
+
+func NewFunctionSelector(pool *model.VectorPool, funcExpr *parser.Call, mint, maxt time.Time, step time.Duration, stepsBatch int, call FunctionCall, majorArgIndex int, majorArg model.VectorOperator, scalarArgs ...model.VectorOperator) model.VectorOperator {
+	return &functionSelector{
+		vectorPool:    pool,
+		mint:          mint.UnixMilli(),
+		maxt:          maxt.UnixMilli(),
+		step:          step.Milliseconds(),
+		currentStep:   mint.UnixMilli(),
+		stepsBatch:    stepsBatch,
+		majorArgIndex: majorArgIndex,
+		majorArg:      majorArg,
+		scalarArgs:    scalarArgs,
+		call:          call,
+		funcExpr:      funcExpr,
+	}
+}
+
+func (o *functionSelector) Explain() (me string, next []model.VectorOperator) {
+	ops := []model.VectorOperator{o.majorArg}
+	ops = append(ops, o.scalarArgs...)
+	return fmt.Sprintf("[*functionSelector] %v(%v)", o.funcExpr.Func.Name, o.funcExpr.Args), ops
+}
+
+func (o *functionSelector) Series(ctx context.Context) ([]labels.Labels, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	return o.series, nil
+}
+
+func (o *functionSelector) GetPool() *model.VectorPool {
+	return o.vectorPool
+}
+
+func (o *functionSelector) Next(ctx context.Context) ([]model.StepVector, error) {
+	if o.currentStep > o.maxt {
+		return nil, nil
+	}
+
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	// Process only non variadic functions.
+	if o.funcExpr.Func.Variadic == 0 {
+		majorInput, err := o.majorArg.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(majorInput) == 0 {
+			return nil, nil
+		}
+
+		scalarInput := make([][]model.StepVector, len(o.scalarArgs))
+		if len(o.scalarArgs) != 0 {
+			for i := range o.scalarArgs {
+				scalarInput[i], err = o.scalarArgs[i].Next(ctx)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		var scalarPoints []promql.Point
+		if len(o.funcExpr.Func.ArgTypes) > 1 && len(scalarInput) > 0 {
+			for j := range scalarInput {
+				scalarPoints = append(scalarPoints, promql.Point{V: scalarInput[j][0].Samples[0]})
+			}
+		}
+
+		switch o.funcExpr.Args[o.majorArgIndex].Type() {
+		case parser.ValueTypeVector, parser.ValueTypeScalar:
```

Comment:
Hm, keep in mind this decision is not made during plan execution, but is really possible to make it at plan compilation because the syntax tree won't change. So Maybe we can precalculate things in the operator constructor.

The next function should be as tight as possible, and basically do
* call next on dowstream
* apply function
* sent result

Reply:
I'll try to modify this!

Reply:
Have modified this to now parse in the operator constructor itself!